### PR TITLE
[WIP] Throwable Accessors

### PR DESCRIPTION
**Pull request by theblixguy**
This is a WIP PR that tracks throwable accessors: https://forums.swift.org/t/throwable-accessors/20509

Checklist:

- [x] Parsing `throws` in accessors
- [x] Type checking `try`/`try!`/`try?` statements
- [x] Type checking protocols requirements and witnesses
- [ ] Type checking edge cases
- [ ] Support for `_read` & `_modify` coroutines
- [x] Type checker tests
- [x] SIL tests

At the moment, the code compiles and works in most scenarios I have tested, but there's still work that needs to be done to ensure this feature is fully ready.

cc @rjmccall @slavapestov

### CODE REVIEW

**Comment by slavapestov** [comment]
Don’t forget to check for throws mismatches in class method overrides as well.

This looks like a good start - I’ll try to take a closer look later, and I’m sure @rjmccall will too!

**Comment by DevAndArtist** [comment]
I think you should also test scenarios of protocol refinement such as the following code snippet but for accessors:

```swift
protocol P {
  func foo() throws
}

protocol Q: P {
  func foo()
}

struct S: Q {
  func foo() {
    print(42)
  }
}
```

I assume if we adapted typed throws the `Q` protocol will make specify a concrete throwing type for `foo` which in that case would be `Never` and therefore can be emitted, this also aligns nicely with accessors.

**Comment by DevAndArtist** [comment]
I updated my example with a few more cases.

**Comment by theblixguy** [comment]
@slavapestov Thanks, I have added a test case for a non-throwing getter/setter being overridden by a throwing one in a class - the diagnostics do need a bit of tweaking so instead of saying "method", it says "accessor".

I have also updated the protocol type checking code to check for `_read` and `_modify` as well, although I couldn't come up with a test case as I can't seem to specify the couroutines explicitly in the requirement. I tested with a class instead, however I think `TypeCheckError` needs to be updated to check for coroutines as I am currently able to override, for example, a non-throwing `_read` with a throwing one in a class.

@DevAndArtist Thanks, that's actually a good example - so this doesn't get caught:

```swift
protocol P1 {
  var foo: Int { get throws }
}

protocol Q1: P1 {
  var foo: Int { get }
}

struct S1: Q1 {
  var foo: Int {
    get { return 0 }
  }
}

let instanceS1 = S1()
let _ = (instanceS1 as P1).foo // No error
```

**Comment by DevAndArtist** [comment]
Also take a look at the example of struct `T` above. It conforms to a protocol that requires the function to throw while the actual implementation does not throw (even the `throws` keyword can be omitted), this is similar to the refinement or overriding examples.

**Comment by theblixguy** [comment]
@DevAndArtist Yep, that passes as well without an error diagnostic. Type checking protocols need a _bit_ more work here I guess.

**Comment by theblixguy** [comment]
@DevAndArtist Fixed.

**Comment by theblixguy** [comment]
I think I have covered most type checking scenarios now @slavapestov @DevAndArtist, let me know if you think these are sufficient and correct (look at the test file for examples).

**Comment by theblixguy** [comment]
@slavapestov @rjmccall I have added some SILGen tests as well. What would be the next steps from here to support `_read` and `_modify` coroutines?

File: lib/Sema/CodeSynthesis.cpp [hunk 1]
```diff
@@ -340,6 +340,10 @@ createCoroutineAccessorPrototype(AbstractStorageDecl *storage,
 
   SourceLoc loc = storage->getLoc();
 
+  AccessorDecl *storageAccessor = nullptr;
+  storageAccessor =
+      kind == AccessorKind::Read ? storage->getGetter() : storage->getSetter();
```
Comment (rjmccall) [inline lib/Sema/CodeSynthesis.cpp]: Hmm.  We should probably be following the logic we use for `mutating`-ness, which is to say that type-checking a storage declaration should compute whether its read/write/read-write operations throw and then store that in the AST, rather than trying to treat the `throws` bit on the accessor as the canonical record of this information.

File: lib/Sema/CodeSynthesis.cpp [hunk 2]
```diff
@@ -826,6 +831,12 @@ static void synthesizeCoroutineAccessorBody(AccessorDecl *accessor,
                                             ASTContext &ctx) {
   assert(accessor->isCoroutine());
 
+  // FIXME: If we don't return here, then a throwing accessor can cause an
+  // infinite loop in SILGen
+  if (!accessor->getStorage()->hasStorage()) {
+    return;
+  }
```
Comment (rjmccall) [inline lib/Sema/CodeSynthesis.cpp]: It's fine if something like this is necessary for incremental development, but I feel like the hack should go in SILGen, like maybe just not emitting a synthetic throwing accessor definition.

File: lib/Sema/TypeCheckDeclOverride.cpp [hunk 3]
```diff
@@ -1672,6 +1672,20 @@ static bool checkSingleOverride(ValueDecl *override, ValueDecl *base) {
...
+      if (auto baseReadCoroutine = baseVD->getReadCoroutine()) {
+        if (overrideReadCoroutine->hasThrows() &&
+            !baseReadCoroutine->hasThrows()) {
+          diags.diagnose(overrideReadCoroutine, diag::override_throws,
+                         isa<ConstructorDecl>(override));
+          diags.diagnose(baseReadCoroutine, diag::overridden_here);
+        }
+      }
+    }
+  }
```
Comment (rjmccall) [inline lib/Sema/TypeCheckDeclOverride.cpp]: Why is this specific to read coroutines?

File: lib/Sema/TypeCheckError.cpp [hunk 4]
```diff
@@ -1647,6 +1674,280 @@ class CheckErrorCoverage : public ErrorHandlingWalker<CheckErrorCoverage> {
...
+                                   contextFlags, typeChecker, tryLoc,
+                                   isTryHandled, true),
+                    nullptr};
+          }
+        }
+
+        return {checkComputedPropertySetter(AE, contextFlags, typeChecker,
+                                            tryLoc, isTryHandled),
+                nullptr};
+      }
```
Comment (rjmccall) [inline lib/Sema/TypeCheckError.cpp]: Yeah, we really need to reintroduce `LValueAccessKind`.  That can be a separate PR.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 5]
```diff
@@ -1313,6 +1313,57 @@ bool WitnessChecker::checkWitnessAccess(ValueDecl *requirement,
...
+            witnessStorageDecl->getModifyCoroutine()) {
+      if (!reqModifyCoroutine->hasThrows() &&
+          witnessModifyCoroutine->hasThrows()) {
+        shouldDiagnose = true;
+      }
+    }
+  }
+
+  return shouldDiagnose;
+}
```
Comment (rjmccall) [inline lib/Sema/TypeCheckProtocol.cpp]: This should all be based on the `doesReadThrow()` etc. bits on the storage declaration.

File: lib/Sema/CodeSynthesis.cpp (hunk 1 above)
Comment (theblixguy) [reply lib/Sema/CodeSynthesis.cpp]: @rjmccall Thanks, I have added two bits to the storage declaration to record whether the getter/setter throws, similar to mutating-ness.

File: lib/Sema/CodeSynthesis.cpp (hunk 2 above)
Comment (theblixguy) [reply lib/Sema/CodeSynthesis.cpp]: @rjmccall Do you mean I should modify `SILGenFunction::emitCoroutineAccessor()` to not emit the couroutine accessor?

File: lib/Sema/TypeCheckDeclOverride.cpp (hunk 3 above)
Comment (theblixguy) [reply lib/Sema/TypeCheckDeclOverride.cpp]: @rjmccall The check is already handled for `_modify`, but I haven't figured out why it doesn't work for `_read` so I had to add this check.

File: lib/Sema/TypeCheckError.cpp (hunk 4 above)
Comment (theblixguy) [reply lib/Sema/TypeCheckError.cpp]: @rjmccall Do we need it to get this PR merged or can it be added later? I have a PR ready to bring back `LValueAccessKind` but it's been a long time since it was removed, so unfortunately the PR breaks a bunch of stuff. Maybe @slavapestov could take a look at bringing it back properly since he removed it?

File: lib/Sema/TypeCheckProtocol.cpp (hunk 5 above)
Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp]: @rjmccall Done!

**Comment by theblixguy** [comment]
I have made some changes based on @rjmccall's feedback.

File: lib/Sema/CodeSynthesis.cpp (hunk 2 above)
Comment (rjmccall) [reply lib/Sema/CodeSynthesis.cpp]: Well, to emit it with a trivial body or something, yes.

File: lib/Sema/TypeCheckError.cpp (hunk 4 above)
Comment (rjmccall) [reply lib/Sema/TypeCheckError.cpp]: I think it's reasonable to ask Slava to help with that, yes.  But I do think we should block this PR on that one instead of adding new logic like this; there's enough of this sort of thing as it is.

Comment (theblixguy) [reply lib/Sema/TypeCheckError.cpp]: @rjmccall Okay! I read Slava's comment on SR-9960 about bringing back LValueAccessKind as some sort of reusable [AST] visitor. If Slava has some free some, then I'll let him do it, otherwise I can pick up the work if someone could point me towards how the computation is supposed to work.

Comment (theblixguy) [reply lib/Sema/TypeCheckError.cpp]: or... I can put up my PR that reverts the LValueAccessKind code that @slavapestov deleted and we can go from there - i.e. fix whatever is broken.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
